core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,278 @@
+
+        Authentication authentication = Jenkins.getAuthentication();
+        String authenticationName = authentication == Jenkins.ANONYMOUS ? "" : authentication.getName();
+
+        try {
+            return new Token(completeUrl, authenticationName, new Date().getTime());
+        } catch (Exception ex) {
+            LOGGER.log(Level.WARNING, "Failed to encode token for URL: " + completeUrl + " user: " + authenticationName, ex);
+        }
+        return null;
+    }
+
+    /**
+     * Implements the browsing support for a specific {@link DirectoryBrowserSupport} like permission check.
+     */
+    private static class InternalResourceRequest {
+        private final String authenticationName;
+        private final String browserUrl;
+
+        InternalResourceRequest(@Nonnull String browserUrl, String authenticationName) {
+            this.browserUrl = browserUrl;
+            this.authenticationName = authenticationName;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            AccessControlled requestRoot = Jenkins.get();
+            String requestUrlSuffix = this.browserUrl;
+
+            LOGGER.fine(() -> "Performing a request as authentication: " + authenticationName + " to object: " + requestRoot + " and restOfUrl: " + requestUrlSuffix + " and restOfPath: " + restOfPath);
+
+            Authentication auth = Jenkins.ANONYMOUS;
+            if (authenticationName != null) {
+                User user = User.getById(authenticationName, false);
+                if (user != null) {
+                    auth = user.impersonate();
+                }
+            }
+
+            try (ACLContext ignored = ACL.as(auth)) {
+                try {
+                    Stapler.getCurrent().invoke(req, rsp, requestRoot, requestUrlSuffix + restOfPath);
+                } catch (Exception ex) {
+                    // cf. UnwrapSecurityExceptionFilter
+                    Throwable cause = ex.getCause();
+                    while (cause != null) {
+                        if (cause instanceof AccessDeniedException) {
+                            throw (AccessDeniedException) cause;
+                        }
+                        cause = cause.getCause();
+                    }
+                    throw ex;
+                }
+                /*
+                While we could just redirect below to the real URL like we do for expired resource URLs, the question
+                is whether we'd end up in a redirect loop if the exception is specific to this mode (and the "normal"
+                URLs redirect to resource URLs). That seems even worse than an error here.
+                 */
+            } catch (AccessDeniedException ade) {
+                LOGGER.log(Level.INFO, "Failed permission check for resource URL access", ade);
```

Comment:
Is this stack trace helpful to admins? Will there be some meaningful details?

Reply:
Going with `FINE`, that combined with the earlier log message on the same level showing exactly what was accessed should be useful enough if something goes wrong here and persists.